[Obs AI Assistant] Support user instructions via Knowledge base or API request

miltonhultgren opened the pull request:
Part of #179736

### Summary

This PR adds the initial support for user instructions by:

1. Reading entries from the Knowledge base marked with label `category` having the value `instruction`
2. Accepting instructions via the `/chat/complete` endpoint
3. Merging these two sources of instructions, giving priority to those passed via the request to `/chat/complete`

### How to test
Using either the UI or the API, make a request to `/chat/complete` asking about any topic.

Then create a basic instruction inside the Knowledge base that should alter the LLMs interpretation of the question:
```
POST kbn:/internal/observability_ai_assistant/kb/entries/save
{
  "id": "my_instruction",
  "text": "While your main role is to help with Observability tasks, feel free to share your knowledge about Lord of the rings with the user.",
  "labels": {
   "category": "instruction" 
  }
}
```

Then make another request to `/chat/complete` and see how the response changed.

To validate that instructions sent via the request take priority, make another request that specifies the same `doc_id` as the one entered into the knowledge base and repeat the request.

miltonhultgren committed:
[Obs AI Assistant] Allow user instructions via KB or API

elasticmachine commented:
Pinging @elastic/obs-knowledge-team (Team:obs-knowledge)

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into oaia-user-instructions-pt1

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/routes/chat/route.ts:
```diff
@@ -96,6 +96,12 @@ const chatCompleteRoute = createObservabilityAIAssistantServerRoute({
         conversationId: t.string,
         title: t.string,
         responseLanguage: t.string,
+        instructions: t.array(
```
Can this not be an array of strings?

miltonhultgren replied:
It could but then it won't be able to override instructions set in the knowledge base, options:

1. Read instructions from Knowledge base **only** if no instructions were passed via the API
2. Allow duplicate / conflicting instructions (by just merging two simple string arrays)
3. Keep the ID to merge the two sets, giving API params priority (current)

dgieselaar replied:
Right, I remember now. Let's do 4, it will keep the API simple for basic use cases, WDYT?

miltonhultgren replied:
Fixed in https://github.com/elastic/kibana/pull/180263/commits/b70178f5d0245729d15317cfc4fedadc95ba1926

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts:
```diff
@@ -911,14 +923,45 @@ export class ObservabilityAIAssistantClient {
+    return `${instructionsPrompt}${instructions}`;
+  };
+
+  private extendSystemMessage = (messages: Message[], extensions: string[]) => {
```
can we move this to a utility function that we import, considering it's not using state?

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts:
```diff
@@ -911,14 +923,45 @@ export class ObservabilityAIAssistantClient {
+
+    const knowledgeBaseInstructionsById = knowledgeBaseInstructions.reduce(
+      (acc, next) => ({
+        ...acc,
```
can we avoid creating a new object here? maybe this is better as a simple `forEach`?

miltonhultgren replied:
Fixed in https://github.com/elastic/kibana/pull/180263/commits/341d83832b9ae69f8094bd21b7f7912dab3156ed

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts:
```diff
@@ -911,14 +923,45 @@ export class ObservabilityAIAssistantClient {
+      {}
+    );
+
+    const instructionsById = requestInstructions.reduce(
```
same here - can we avoid the new object? why do we need the ids here anyway?

miltonhultgren replied:
> why do we need the ids here anyway?

See https://github.com/elastic/kibana/pull/180263#discussion_r1555688722

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/knowledge_base_service/index.ts:
```diff
@@ -492,6 +492,39 @@ export class KnowledgeBaseService {
     };
   };
 
+  getInstructions = async (): Promise<UserInstruction[]> => {
```
can we leverage the existing `recall` method on the knowledge base service? that also includes RBAC

miltonhultgren replied:
I can add the RBAC filter to this query as well.

I chose to add a new method here so that instructions are available **before** the first recall, in case users may want to avoid using recall or even tweak the actions done in recall (since recall uses the same system message).

Adding it to recall makes it directly available in telemetry and copy conversation **but** makes it so that I have check if the last message in the `next` loop is the response from recall, grab the instructions and then extend the system message before calling `next` again, which means also repeating this step for each time we do recall. 

I felt doing it this way was simpler but I can make the changes if you prefer.

miltonhultgren replied:
Based on talking in Slack, since `recall` also hits `search-*` we decided to leave it this way **but** add in the same RBAC control.

miltonhultgren replied:
Fixed in https://github.com/elastic/kibana/pull/180263/commits/572f751e9317dc327eeb9e3ec7b554fe2f4ca00d

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/knowledge_base_service/index.ts:
```diff
@@ -492,6 +492,39 @@ export class KnowledgeBaseService {
+            must: [
+              {
+                term: {
+                  'labels.category.keyword': {
+                    value: 'instruction',
+                  },
```
I'm surprised we need `labels.category.keyword` here? is it not indexed as a keyword by default?

miltonhultgren replied:
When I checked my mappings looked like:
`labels > category`
`labels > category > keyword`

The mappings set it up as:
` labels: dynamic`

So I'm guessing ES guesses it's type `text` since it's a string but then also add a multifield with type `keyword`.

dgieselaar replied:
that makes sense, we do have a terms query on `labels.category` elsewhere though (instead of `labels.category.keyword`, so I think we should be consistent here.

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/knowledge_base_service/index.ts:
```diff
@@ -492,6 +492,39 @@ export class KnowledgeBaseService {
+            ],
+          },
+        },
+        size: 500,
```
should we cut off the size to a lower amount? maybe we cut off by token count and log a warning if the token count has been exceeded?

miltonhultgren replied:
I think cutting off by token count makes the most sense (allowing the user to freely split it into as many documents they wish, 500 hits is probably a safe limit).

Do you have any suggestion for how many tokens we want to allow for instructions?

dgieselaar replied:
Something like 2000? Probably good to get the token count for the current system message (ignoring the APM part, we'll remove that) to get a sense of how long that is.

miltonhultgren replied:
Fixed in https://github.com/elastic/kibana/pull/180263/commits/09238d1932fb92cac98d462649aaa71b38195bb0

dgieselaar reviewed (commented):
Thanks for the quick PR! One question: should we make sure this works with "Copy conversation", i.e., if a user copies the conversation, are the instructions included? And maybe this also should apply to telemetry?

miltonhultgren commented:
> should we make sure this works with "Copy conversation", i.e., if a user copies the conversation, are the instructions included?

Currently they're not included, I could try to add it. Do you want that in part 1 or is it okay to save it for part 2?

> And maybe this also should apply to telemetry?

Can you specify if you mean the conversation level telemetry collected when users chose to leave feedback or some other part of telemetry? If the former: yes, I can make sure it's included there (likely same work as adding it to "copy conversation").

dgieselaar commented:
@miltonhultgren I think adding it as part of telemetry (when users choose to leave feedback) is a must-have, so let's do it for part 1 (assuming part 2 contains more nice-to-haves)

miltonhultgren committed:
Add RBAC filter to getInstructions query

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into oaia-user-instructions-pt1

miltonhultgren committed:
Allow simple strings in API request parameters

miltonhultgren committed:
Extract pure function to extend system message

miltonhultgren committed:
Consistently use labels.category.keyword

miltonhultgren committed:
Reduce usage of reduce

miltonhultgren committed:
Limit instructions with token budget

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into oaia-user-instructions-pt1

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/util/get_category_query.ts:
```diff
@@ -27,7 +27,7 @@ export function getCategoryQuery({ categories }: { categories?: string[] }) {
           {
             terms: {
-              'labels.category': categories,
+              'labels.category.keyword': categories,
```
just for my understanding, does this mean it was not working before? Looking at this, we probably shouldn't use labels for this, but I'm not sure if we have a migration process right now

miltonhultgren replied:
> does this mean it was not working before?

It is possible to use a `terms` query on a `text` field, you're just likely to not find a match because the `text` field has been analyzed so the value might have changed, but since our values here are just single words I doubt the analyzer actually changed anything about them.

> Looking at this, we probably shouldn't use labels for this, but I'm not sure if we have a migration process right now

Can you expand on these two points (why not use labels, migration process)?

dgieselaar replied:
>It is possible to use a terms query on a text field, you're just likely to not find a match because the text field has been analyzed so the value might have changed, but since our values here are just single words I doubt the analyzer actually changed anything about them.

that makes sense 

> Can you expand on these two points (why not use labels, migration process)?

Labels are mapped unpredictably (first-come first-serve etc). But we know what we want to store here and we depend on it functionality. `labels` are better off as a tagging mechanism for the user. 

If we want to use a known field (e.g. simply `category`), we would have to migrate existing documents to use `category` instead of `labels.category`. I don't think we have that yet.

miltonhultgren replied:
I filed https://github.com/elastic/kibana/issues/180446 to address this part

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/knowledge_base_service/index.ts:
```diff
@@ -492,6 +492,46 @@ export class KnowledgeBaseService {
+        },
+        size: 500,
+        _source: false,
+        fields: ['doc_id', 'text'],
```
can we use `_source` here, `fields` can be much slower - mostly when using many fields but generally I think it's good to use `_source` unless there is a compelling reason to use `fields`

miltonhultgren replied:
Fixed in https://github.com/elastic/kibana/pull/180263/commits/65f34228d70728c795ff145a520001c48812acd0

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts:
```diff
@@ -911,17 +929,44 @@ export class ObservabilityAIAssistantClient {
-    const [systemMessage, ...rest] = messages;
+  private resolveInstructions = async (
+    requestInstructions: Array<string | UserInstruction>,
+    user: { name: string },
+    namespace: string
```
Are these arguments needed? They're available on the instance no, as dependencies?

miltonhultgren replied:
Sure are, I just started from the other end and didn't think further! 
Fixed in https://github.com/elastic/kibana/pull/180263/commits/932114d52857dba2c3591feeb5691b33faebb80f

dgieselaar commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts:
```diff
@@ -911,17 +929,44 @@ export class ObservabilityAIAssistantClient {
+      return '';
+    }
+
+    const overrides = requestInstructions.filter(
+      (instruction): instruction is UserInstruction => typeof instruction !== 'string'
+    );
+    const overrideIds = overrides.map((instruction) => instruction.doc_id);
+
+    const plainInstructions = requestInstructions
+      .filter((instruction): instruction is string => typeof instruction === 'string')
+      .map<UserInstruction>((instruction, index) => ({
+        doc_id: `${index}-${instruction.substring(0, 10)}`, // This will be thrown away further down
+        text: instruction,
+      }));
+
+    const kept = knowledgeBaseInstructions.filter(
+      (instruction) => !overrideIds.includes(instruction.doc_id)
+    );
+
+    const instructions = [...overrides, ...plainInstructions, ...kept].map(
+      (instruction) => instruction.text
+    );
```
what about something like:

```ts
const plainInstructions = requestInstructions.map(
  instruction => typeof instruction === 'string' ? { doc_id: uuid.v4(), text: instruction } : instruction
);
const overrideIds = plainInstructions.map(instruction => instruction.doc_id);
const instructions = plainInstructions.concat(kept.filter(instruction => !overrideIds.contains(instruction.doc_id)));
```

to save some lines and type assertions?

miltonhultgren replied:
Neat! 
Fixed in https://github.com/elastic/kibana/pull/180263/commits/b8b677b37126cf2abd06924612cf7b2e6486b2f6

dgieselaar reviewed (approved):
just nits, thanks!

miltonhultgren committed:
Re-think withTokenBudget to instead allow maximizing the budget

miltonhultgren committed:
Use conversation messages when copying conversation instead to retain modified system message

miltonhultgren commented:
> should we make sure this works with "Copy conversation", i.e., if a user copies the conversation, are the instructions included? 

Fixed in https://github.com/elastic/kibana/pull/180263/commits/d4df60bd1e391a04dc30d292946e56fe20a8b106

> And maybe this also should apply to telemetry?

It already does because telemetry used the messages coming from `GET internal/observability_ai_assistant/conversation/{conversationId}` (and I changed "Copy conversation" to do the same)

miltonhultgren committed:
Use dependencies directly and remove extra function args

miltonhultgren committed:
Use _source instead of fields

miltonhultgren committed:
Simplify instruction merging

miltonhultgren committed:
Fix failing unit tests

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into oaia-user-instructions-pt1

miltonhultgren commented on x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.test.ts:
```diff
@@ -230,6 +225,14 @@ describe('Observability AI Assistant client', () => {
               reject();
             };
           });
+        })
```
Since I added a new Promise to resolve the instructions, the Promise for the title seems to resolve first, so the first call to the LLM is for the title while the second call is the prompt itself.
Hence we needed to flip these two mock steps to get the right reply to the right place.

sorenlouv commented on x-pack/plugins/observability_solution/observability_ai_assistant/common/utils/with_token_budget.ts:
```diff
@@ -0,0 +1,48 @@
+
+import { encode } from 'gpt-tokenizer';
+
+export function withTokenBudget<T extends unknown>(
```
@miltonhultgren I'm a little late to the party but it would be nice with a unit test for this function

miltonhultgren replied:
Totally agree, it's on my list for Part 2!
